diff -urN o/gen_wrap.py n/gen_wrap.py
--- o/gen_wrap.py	1970-01-01 03:00:00.000000000 +0300
+++ n/gen_wrap.py	2021-06-13 09:26:34.111773467 +0300
@@ -1448,8 +1448,8 @@
         fdata.read_header("barvinok/isl.h")
 
     for part, classes in PART_TO_CLASSES.items():
-        expf = open("src/wrapper/gen-expose-%s.inc" % part, "wt")
-        wrapf = open("src/wrapper/gen-wrap-%s.inc" % part, "wt")
+        expf = open("src/wrapper/gen-expose-%s.inc" % part, "wb")
+        wrapf = open("src/wrapper/gen-wrap-%s.inc" % part, "wb")
 
         classes = [
                 cls
diff -urN o/isl/isl_hash.c n/isl/isl_hash.c
--- o/isl/isl_hash.c	2020-07-02 00:00:10.000000000 +0300
+++ n/isl/isl_hash.c	2021-06-13 09:45:26.570769147 +0300
@@ -8,6 +8,7 @@
  */
 
 #include <stdlib.h>
+#include <strings.h>
 #include <isl/hash.h>
 #include <isl/ctx.h>
 #include "isl_config.h"
diff -urN o/setup.py n/setup.py
--- o/setup.py	1970-01-01 03:00:00.000000000 +0300
+++ n/setup.py	2021-06-13 09:24:02.719774045 +0300
@@ -196,11 +196,7 @@
             if "isl_ast_int.c" in fn and conf["USE_SHIPPED_IMATH"]:
                 continue
 
-            inf = open(fn, "r", encoding="utf-8")
-            try:
-                contents = inf.read()
-            finally:
-                inf.close()
+            contents = open(fn, "rb").read()
 
             if "int main(" not in contents and not blacklisted:
                 EXTRA_OBJECTS.append(fn)
